src/taipy/core/pipeline/_pipeline_manager.py

```diff
@@ -90,42 +108,96 @@ def __remove_subscriber(cls, callback, params, pipeline):
-            owner_id,
-            {scenario_id} if scenario_id else None,
+            properties=properties,
+            tasks=_tasks,
+            pipeline_id=pipeline_id,
+            owner_id=scenario_id,
+            parent_ids={scenario_id} if scenario_id else None,
+            subscribers=subscribers,
             version=version,
         )
-        for task in tasks:
-            task._parent_ids.update([pipeline_id])
-        cls.__save_tasks(tasks)
-        cls._set(pipeline)
+        for task in _tasks:
+            if pipeline_id not in task._parent_ids:
+                task._parent_ids.update([pipeline_id])
+                task_manager._set(task)
         _publish_event(cls._EVENT_ENTITY_TYPE, pipeline.id, EventOperation.CREATION, None)
         return pipeline
 
     @classmethod
-    def __save_tasks(cls, tasks):
-        task_manager = _TaskManagerFactory._build_manager()
-        for i in tasks:
-            task_manager._set(i)
+    def _get(cls, pipeline: Union[str, Pipeline], default=None) -> Pipeline:
+        """
+        Returns a Pipeline by id or reference.
+        """
+        try:
+            pipeline_id = pipeline.id if isinstance(pipeline, Pipeline) else pipeline
+            pipeline_name, scenario_id = cls._breakdown_pipeline_id(pipeline_id)
+
+            scenario_manager = _ScenarioManagerFactory._build_manager()
+            if scenario := scenario_manager._get(scenario_id):
+                if pipeline_entity := scenario.pipelines.get(pipeline_name, None):
+                    return pipeline_entity
+            cls.__log_error_entity_not_found(pipeline_id)
+            return default
+        except (ModelNotFound, InvalidPipelineId):
+            cls.__log_error_entity_not_found(pipeline_id)
+            return default
+
+    @classmethod
+    def __log_error_entity_not_found(cls, pipeline_id: Union[PipelineId, str]):
+        cls._logger.error(f"{cls._ENTITY_NAME} not found: {str(pipeline_id)}")
+
+    @classmethod
+    def _set(cls, pipeline: Pipeline):
+        """
+        Save or update a Pipeline.
+        """
+        pipeline_name, scenario_id = cls._breakdown_pipeline_id(pipeline.id)
+        scenario_manager = _ScenarioManagerFactory._build_manager()
+
+        if scenario := scenario_manager._get(scenario_id):
+            pipeline_data = {
+                Scenario._PIPELINE_TASKS_KEY: pipeline._tasks,
+                Scenario._PIPELINE_SUBSCRIBERS_KEY: pipeline._subscribers,
+                Scenario._PIPELINE_PROPERTIES_KEY: pipeline._properties.data,
+            }
+            scenario._pipelines[pipeline_name] = pipeline_data
+            scenario_manager._set(scenario)
+        else:
+            cls._logger.error(f"Pipeline {pipeline.id} belongs to a non-existing Scenario {scenario_id}.")
+            raise PipelineBelongsToNonExistingScenario(pipeline.id, scenario_id)
+
+    @classmethod
+    def _breakdown_pipeline_id(cls, pipeline_id: str) -> Tuple[str, str]:
```

Comment:
```suggestion
    def __breakdown_pipeline_id(cls, pipeline_id: str) -> Tuple[str, str]:
```
I think this should be a private method

Reply:
Actually kept is as _breakdown_pipeline_id so that I can import it for testing 😅  but let me see if we can do that and mypy will still be happy

Reply:
that s fine to keep it with a single _